s2/gen.go

```diff
@@ -0,0 +1,1590 @@
+		JMP(LabelRef(name + "tail"))
+	*/
+
+	Label(name + "move_1or2")
+	MOVB(Mem{Base: src}, AX.As8())
+	MOVB(Mem{Base: src, Disp: -1, Index: length, Scale: 1}, CX.As8())
+	MOVB(AX.As8(), Mem{Base: dst})
+	MOVB(CX.As8(), Mem{Base: dst, Disp: -1, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_4")
+	MOVL(Mem{Base: src}, AX.As32())
+	MOVL(AX.As32(), Mem{Base: dst})
+	JMP(end)
+
+	Label(name + "move_3")
+	MOVW(Mem{Base: src}, AX.As16())
+	MOVB(Mem{Base: src, Disp: 2}, CX.As8())
+	MOVW(AX.As16(), Mem{Base: dst})
+	MOVB(CX.As8(), Mem{Base: dst, Disp: 2})
+	JMP(end)
+
+	Label(name + "move_5through7")
+	MOVL(Mem{Base: src}, AX.As32())
+	MOVL(Mem{Base: src, Disp: -4, Index: length, Scale: 1}, CX.As32())
+	MOVL(AX.As32(), Mem{Base: dst})
+	MOVL(CX.As32(), Mem{Base: dst, Disp: -4, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_8")
+	// We need a separate case for 8 to make sure we write pointers atomically.
+	MOVQ(Mem{Base: src}, AX)
+	MOVQ(AX, Mem{Base: dst})
+	JMP(end)
+
+	Label(name + "move_9through16")
+	MOVQ(Mem{Base: src}, AX)
+	MOVQ(Mem{Base: src, Disp: -8, Index: length, Scale: 1}, CX)
+	MOVQ(AX, Mem{Base: dst})
+	MOVQ(CX, Mem{Base: dst, Disp: -8, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_17through32")
+	X0, X1, X2, X3, X4, X5, X6, X7 := XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM()
+	X8, X9, X10, X11, X12, X13, X14, X15 := XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM()
+
+	MOVOU(Mem{Base: src}, X0)
+	MOVOU(Mem{Base: src, Disp: -16, Index: length, Scale: 1}, X1)
+	MOVOU(X0, Mem{Base: dst})
+	MOVOU(X1, Mem{Base: dst, Disp: -16, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_33through64")
+	MOVOU(Mem{Base: src}, X0)
+	MOVOU(Mem{Base: src, Disp: 16}, X1)
+	MOVOU(Mem{Base: src, Disp: -32, Index: length, Scale: 1}, X2)
+	MOVOU(Mem{Base: src, Disp: -16, Index: length, Scale: 1}, X3)
+	MOVOU(X0, Mem{Base: dst})
+	MOVOU(X1, Mem{Base: dst, Disp: 16})
+	MOVOU(X2, Mem{Base: dst, Disp: -32, Index: length, Scale: 1})
+	MOVOU(X3, Mem{Base: dst, Disp: -16, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_65through128")
+	MOVOU(Mem{Base: src}, X0)
+	MOVOU(Mem{Base: src, Disp: 16}, X1)
+	MOVOU(Mem{Base: src, Disp: 32}, X2)
+	MOVOU(Mem{Base: src, Disp: 48}, X3)
+	MOVOU(Mem{Base: src, Index: length, Scale: 1, Disp: -64}, X12)
+	MOVOU(Mem{Base: src, Index: length, Scale: 1, Disp: -48}, X13)
+	MOVOU(Mem{Base: src, Index: length, Scale: 1, Disp: -32}, X14)
+	MOVOU(Mem{Base: src, Index: length, Scale: 1, Disp: -16}, X15)
+	MOVOU(X0, Mem{Base: dst})
+	MOVOU(X1, Mem{Base: dst, Disp: 16})
+	MOVOU(X2, Mem{Base: dst, Disp: 32})
+	MOVOU(X3, Mem{Base: dst, Disp: 48})
+	MOVOU(X12, Mem{Base: dst, Index: length, Scale: 1, Disp: -64})
+	MOVOU(X13, Mem{Base: dst, Index: length, Scale: 1, Disp: -48})
+	MOVOU(X14, Mem{Base: dst, Index: length, Scale: 1, Disp: -32})
+	MOVOU(X15, Mem{Base: dst, Index: length, Scale: 1, Disp: -16})
+	JMP(end)
+
+	Label(name + "move_129through256")
+	MOVOU(Mem{Base: src}, X0)
```

Comment:
If the `X*` registers were an array, this whole block of instructions could be loops instead. This applies elsewhere too.

I'm guessing this was a direct port from the standard library?

Reply:
Indeed.